### ISSUE TITLE
CosineLRwWarmup

### ISSUE THREAD

**Issue by anishacharya** [lightly-ai/lightly#1352]
This code is wrong when lr0 != 1

https://github.com/lightly-ai/lightly/blob/b69b8b14c29121422479f23078488efca734a995/lightly/utils/scheduler.py#L45

since you are returning decay to superclass lambdaLR the resulting 
`lr = decay * start_value`

example when  the cosine schedule kicks off i.e. step=0 
```
decay = ( 
end_value - (end_value - start_value)* (np.cos(np.pi * step / (max_steps - 1)) + 1) / 2
) 
= start_value
```

which should be the learning rate not decay 

Also typing is wrong:
should be:
```
def __init__(
			self,
			optimizer: torch.optim.Optimizer,
			warmup_epochs: int,
			max_epochs: int,
			last_epoch: int = -1,
			start_value: float = 1.0,
			end_value: float = 0.001,
			verbose: bool = False,
	) -> None:
```

epochs should not be float. 

One easy fix if you want to stick with the LambdaLR logic is to make the cosine rate return: decay / start_value 
then LambdaLR will make lr = decay

**Comment by anishacharya** [comment]
Here is a quick fix 
https://github.com/lightly-ai/lightly/pull/1353

Hope this helps.

**Comment by philippmwirth** [comment]
Hi @anishacharya, I think the code you linked is correct. We use `cosine_schedule` as the `lr_lambda` we pass to `LambdaLR`. From the docs of `LambdaLR` ([link](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html)):
> lr_lambda (function or [list](https://docs.python.org/3/library/stdtypes.html#list)) – A function which computes a multiplicative factor given an integer parameter epoch, or a list of such functions, one for each group in optimizer.param_groups.

So the function is expected to return the decay, not the learning rate.

You are right about the typehints though, thanks!

**Comment by anishacharya** [comment]
That is right. However, I think the current implementation is wrongly returning the lr instead of decay ~ can you see the PR I linked ? I was getting wrong learning rate values when using the current lightly implementation ~ i added a simple test case to verify

**Comment by philippmwirth** [comment]
I'll take the conversation to the linked PR 👍

**Comment by guarin** [comment]
Done in #1353

**Comment by anishacharya** [comment]
The issue persists....
@guarin @philippmwirth @IgorSusmelj 
The issue was not in your warmup -- the issue is in your decay.... Please note that you should be returning "scaling factor" as  you are using torch.optim.lr_scheduler.LambdaLR 

so an easy fix is:

```
class CosineWarmupScheduler(torch.optim.lr_scheduler.LambdaLR):
	"""
    Cosine warmup scheduler for learning rate.
    Args:
        optimizer:
            Optimizer object to schedule the learning rate.
        warmup_epochs:
            Number of warmup epochs or steps.
        max_epochs:
            Total number of training epochs or steps.
        last_epoch:
            The index of last epoch or step. Default: -1
        start_value:
            Starting learning rate scale. Default: 1.0
        end_value:
            Target learning rate scale. Default: 0.001
        verbose:
            If True, prints a message to stdout for each update. Default: False.
    """

	def __init__(
			self,
			optimizer: torch.optim.Optimizer,
			warmup_epochs: int,
			max_epochs: int,
			last_epoch: int = -1,
			start_value: float = 1.0,
			end_value: float = 0.001,
			verbose: bool = False,
	) -> None:
		self.warmup_epochs = warmup_epochs
		self.max_epochs = max_epochs
		self.start_value = start_value
		self.end_value = end_value
		super().__init__(
			optimizer=optimizer,
			lr_lambda=self.scale_lr,
			last_epoch=last_epoch,
			verbose=verbose,
		)

	def cosine_schedule(self, step: int, max_steps: int) -> float:
		"""
		Use cosine decay to gradually modify start_value to reach target end_value during iterations.
		@param step
		@param max_steps
		Returns: Cosine decay value.
		"""
		if step < 0:
			raise ValueError("Current step number can't be negative")
		if max_steps < 1:
			raise ValueError("Total step number must be >= 1")
		if step > max_steps:
			warnings.warn(
				f"Current step number {step} exceeds max_steps {max_steps}.",
				category=RuntimeWarning,
			)
		if max_steps == 1:
			# Avoid division by zero
			decay = self.end_value
		elif step == max_steps:
			# Special case for Pytorch Lightning which updates LR scheduler also for epoch
			# after last training epoch.
			decay = self.end_value
		else:
			decay = (
					self.end_value
					- (self.end_value - self.start_value)
					* (np.cos(np.pi * step / (max_steps - 1)) + 1)
					/ 2
			)
		**return decay / self.start_value**. <----- This line changed 

	def scale_lr(self, epoch: int) -> float:
		"""
        Scale learning rate according to the current epoch number.
        Args:
            epoch:
                Current epoch number.
        Returns:
            Scaled learning rate scaling:
        """
		if epoch < self.warmup_epochs:
			**decay = (epoch + 1) / self.warmup_epochs**. < ---- Revert back to this
		else:
			decay = self.cosine_schedule(
				step=epoch - self.warmup_epochs,
				max_steps=self.max_epochs - self.warmup_epochs
			)
		return decay 
```
